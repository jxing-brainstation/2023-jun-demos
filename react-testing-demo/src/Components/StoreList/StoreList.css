.sale {
    border: 5px dotted red;
}