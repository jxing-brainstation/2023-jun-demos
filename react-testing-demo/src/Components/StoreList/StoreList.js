import React from 'react'
import StoreCard from '../StoreCard/StoreCard';
import './StoreList.css'

export default function StoreList({ stores, sale, loadData }) {
    const getMoreStores = () => {
        console.log("Hello world");
        loadData();
    };

    return (
        <div className={sale ? "sale" : ""}>
            <h2>All the Stores!!!</h2>

            {stores.map((item) => (
                <StoreCard key={item.id} address={item.address} />
            ))}

            <button onClick={getMoreStores}>load more stores</button>
        </div>
    );
}
