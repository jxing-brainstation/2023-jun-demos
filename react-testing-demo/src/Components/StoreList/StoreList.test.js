import StoreList from "./StoreList";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const mockStores = [
    {
        id: "1",
        address: "store 1 address"
    },
    {
        id: "2",
        address: "store 2 address"
    },
    {
        id: "3",
        address: "store 3 address"
    }
]


describe("StoreList", () => {
    it("renders a list of StoreCards based on the stores prop", () => {
        //step 1: render the component
        render(<StoreList stores={mockStores} sale={true} loadData={() => { }} />)
        //step 2: write the assertion
        const store1Address = screen.getByText("store 1 address");
        expect(store1Address).toBeTruthy();
        const store2Address = screen.getByText("store 2 address");
        expect(store2Address).toBeTruthy();
        const store3Address = screen.getByText("store 3 address");
        expect(store3Address).toBeTruthy();
    })

    it("applies the sale class based on the sale prop to the container div", () => {
        const { container } = render(<StoreList stores={mockStores} sale={true} loadData={() => { }} />)

        // console.log(container.firstChild)
        expect(container.firstChild.classList.contains("sale")).toBe(true);

    })

    it("renders the load more stores button", () => {
        //Render the component
        render(<StoreList stores={mockStores} sale={true} loadData={() => { }} />)

        //write the assertion
        //get the button from the screen
        //check that it is there
        const button = screen.getByRole("button");
        expect(button).toBeInTheDocument();
    })

    it("calls loadData when the load more stores button is clicked", () => {
        // make a mock of the function we pass as the loadData prop
        const mockLoadData = jest.fn();

        render(<StoreList stores={mockStores} sale={true} loadData={mockLoadData} />)

        // stimulate the user click on the button
        const loadMoreStoresButton = screen.getByText("load more stores");
        userEvent.click(loadMoreStoresButton);

        //write the assertion
        expect(mockLoadData).toBeCalled();
    })

    it("matches the snapshot", () => {
        const view = render(<StoreList stores={mockStores} sale={true} loadData={() => { }} />)

        expect(view).toMatchSnapshot();
    })
})