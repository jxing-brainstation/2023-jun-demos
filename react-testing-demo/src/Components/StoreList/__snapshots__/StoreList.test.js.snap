// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreList component matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sale"
      >
        <h2>
          All the Stores!!!
        </h2>
        <div
          class="store-card"
        >
          store 1 address
        </div>
        <div
          class="store-card"
        >
          store 2 address
        </div>
        <div
          class="store-card"
        >
          store 3 address
        </div>
        <button>
          load more stores
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sale"
    >
      <h2>
        All the Stores!!!
      </h2>
      <div
        class="store-card"
      >
        store 1 address
      </div>
      <div
        class="store-card"
      >
        store 2 address
      </div>
      <div
        class="store-card"
      >
        store 3 address
      </div>
      <button>
        load more stores
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
